### ISSUE TITLE
AI Provider System Cleanup

### ISSUE THREAD

**Issue by rpnunez** [rpnunez/wp-ai-scheduler#1817]
# Plan: Systematic JSON Schema Adoption & Prompt Cleanup

## Context

The plugin was originally built around the Meow Apps AI Engine, whose `simpleJsonQuery()` had reliability issues that led developers to fall back to `generate_text()` + manual JSON parsing. A lot of prompt builders grew "JSON shape descriptions" inline in the prompt text as a workaround.

The WP AI Client (Gemini) now supports native structured JSON via `as_json_response($schema)`, which enforces the response shape at the API level. A `json_schema` option exists in `prepare_options()` (class-aips-ai-service.php:777) and flows through to the provider. We proved this works in `class-aips-research-service.php`.

**Goal:** Apply the `json_schema` pattern everywhere we call `generate_json()`, switch any `generate_text()` calls that expect JSON, and strip verbose JSON shape instructions from prompts (replacing them with concise field hints that still serve the text-based fallback path).

---

### 1. Add `json_schema` to all `generate_json()` callers

Each class that calls `generate_json()` gets a private `get_*_json_schema()` method (following the `class-aips-research-service.php` pattern), and the schema is passed in the options array.

#### `class-aips-author-topics-generator.php` (line 107)
```php
$response = $this->ai_service->generate_json($prompt, array(
    'temperature' => 0.7,
    'json_schema' => $this->get_topic_json_schema(),
));
```
Schema (matches `parse_json_topics()` validation):
```php
private function get_topic_json_schema(): array {
    return array(
        'type'  => 'array',
        'items' => array(
            'type'       => 'object',
            'properties' => array(
                'title'    => array('type' => 'string'),
                'score'    => array('type' => 'integer'),
                'keywords' => array('type' => 'array', 'items' => array('type' => 'string')),
            ),
            'required' => array('title', 'score', 'keywords'),
        ),
    );
}
```

#### `class-aips-author-suggestions-service.php` (line 133)
```php
$response = $this->ai_service->generate_json($prompt, array(
    'temperature' => 0.8,
    'json_schema' => $this->get_author_json_schema(),
));
```
Schema (12 fields, matches `parse_suggestions()` validation; enum values for `expertise_level` and `preferred_content_length` enforce what `normalize_*()` already maps):
```php
private function get_author_json_schema(): array {
    return array(
        'type'  => 'array',
        'items' => array(
            'type'       => 'object',
            'properties' => array(
                'name'                     => array('type' => 'string'),
                'field_niche'              => array('type' => 'string'),
                'description'              => array('type' => 'string'),
                'details'                  => array('type' => 'string'),
                'keywords'                 => array('type' => 'string'),
                'voice_tone'               => array('type' => 'string'),
                'writing_style'            => array('type' => 'string'),
                'target_audience'          => array('type' => 'string'),
                'expertise_level'          => array('type' => 'string', 'enum' => array('beginner', 'intermediate', 'expert', 'thought_leader')),
                'content_goals'            => array('type' => 'string'),
                'preferred_content_length' => array('type' => 'string', 'enum' => array('short', 'medium', 'long')),
                'topic_generation_prompt'  => array('type' => 'string'),
            ),
            'required' => array('name', 'field_niche', 'description', 'details', 'keywords', 'voice_tone', 'writing_style', 'target_audience', 'expertise_level', 'content_goals', 'preferred_content_length', 'topic_generation_prompt'),
        ),
    );
}
```
Note: `keywords` stays `string` (comma-separated) to match how `parse_suggestions()` consumes it.

#### `class-aips-content-auditor.php` (line 118)
```php
$response = $this->ai_service->generate_json($prompt, array(
    'temperature' => 0.7,
    'json_schema' => $this->get_gap_analysis_json_schema(),
));
```
Schema:
```php
private function get_gap_analysis_json_schema(): array {
    return array(
        'type'  => 'array',
        'items' => array(
            'type'       => 'object',
            'properties' => array(
                'missing_topic' => array('type' => 'string'),
                'priority'      => array('type' => 'string', 'enum' => array('High', 'Medium')),
                'reason'        => array('type' => 'string'),
                'search_intent' => array('type' => 'string'),
            ),
            'required' => array('missing_topic', 'priority', 'reason', 'search_intent'),
        ),
    );
}
```

#### `class-aips-campaigns-controller.php` (line 594)
Campaign returns a **single object** (not an array). Current code unwraps `$response[0]` if response is an array-of-arrays — with a schema enforcing an object, that unwrapping becomes unnecessary.
```php
$response = $this->ai_service->generate_json($prompt, array(
    'json_schema' => $this->get_campaign_json_schema(),
));
```
Schema (object, not array):
```php
private function get_campaign_json_schema(): array {
    return array(
        'type'       => 'object',
        'properties' => array(
            'campaign_name'        => array('type' => 'string'),
            'content_goal'         => array('type' => 'string'),
            'post_type'            => array('type' => 'string'),
            'prompt_template'      => array('type' => 'string'),
            'title_prompt'         => array('type' => 'string'),
            'author_persona'       => array('type' => 'string'),
            'campaign_mode'        => array('type' => 'string', 'enum' => array('template', 'author')),
            'review_policy'        => array('type' => 'string', 'enum' => array('draft', 'approval', 'auto_publish')),
            'frequency'            => array('type' => 'string'),
            'time_window_start'    => array('type' => 'string'),
            'time_window_end'      => array('type' => 'string'),
            'post_tags'            => array('type' => 'string'),
            'post_category'        => array('type' => 'integer'),
            'template_style'       => array('type' => 'string'),
            'sample_article_ideas' => array('type' => 'array', 'items' => array('type' => 'string')),
            'risks_assumptions'    => array('type' => 'array', 'items' => array('type' => 'string')),
        ),
        'required' => array('campaign_name', 'content_goal', 'post_type', 'prompt_template', 'title_prompt', 'campaign_mode', 'review_policy', 'sample_article_ideas'),
    );
}
```
Also remove the `$response[0]` unwrap guard in `ajax_ai_generate_campaign()` since the schema guarantees an object response.

---

#### `class-aips-dev-tools.php` (line 91)
Currently calls `generate_text()` then manually strips markdown and `json_decode()`s the result. The scaffold type is known before the call (passed via POST `scaffold_type`), so the schema can be selected at call time.

Switch to `generate_json()` with schema selected by `$scaffold_type`:
```php
$response = $ai_service->generate_json($prompt, array(
    'temperature' => 0.7,
    'json_schema' => $this->get_scaffold_json_schema($scaffold_type),
));
```
Remove the manual `json_decode` + markdown-stripping block (lines 98-110 approx) — `generate_json()` already handles that internally.

Three scaffold schemas (one per `$scaffold_type`): `voice`, `article_structure`, `template`. Each is a single object with the field set the prompt already describes.

---

### 3. Prompt builder cleanup

**Strategy:** Remove verbose "Example format:" blocks and "Return ONLY a valid JSON array / Return a JSON array..." paragraphs. Replace with a single concise line that serves as fallback guidance. This keeps the prompt readable for the Meow/text-fallback path while eliminating prompt-bloat for native JSON.

Pattern to apply across all prompt builders that request structured JSON:

**Before (verbose):**
```
Return a JSON array of objects. Each object must have:
- "title": The blog post topic/title (string)
- "score": Estimated engagement score 1-100 (integer)
- "keywords": 3-5 relevant keywords (array of strings)

Example format:
[
  {
    "title": "10 Best Practices...",
    "score": 85,
    "keywords": ["WordPress", "SEO", ...]
  }
]
```

**After (concise):**
```
Return a JSON array where each item has: "title" (string), "score" (integer 1-100), "keywords" (string array).
```

Files to update:
- `class-aips-prompt-builder-topic.php` (lines 191-203) — remove example block, replace with one-liner
- `class-aips-prompt-builder-authors.php` (lines 95-125) — remove example block, replace with one-liner
- `class-aips-research-service.php` `build_research_prompt()` (lines 133-149) — same
- `class-aips-research-service.php` `build_source_research_prompt()` (lines 477-482) — same
- `class-aips-content-auditor.php` (wherever the gap analysis prompt is built)
- `class-aips-campaigns-controller.php` (the campaign generation prompt, line ~540-590)
- `class-aips-dev-tools.php` (the scaffold prompt, the "Return ONLY the raw JSON object" line)

---

### 4. Out of scope for this pass

- `class-aips-taxonomy-controller.php` — uses `generate_text()` with plain line-by-line output. Simple format, no JSON. Not worth switching; the current approach is correct for this use case.
- `class-aips-seeder-service.php` — uses `generator->generate_content()` (text path) internally. Seeder is a dev/setup utility; lower risk to leave as-is.
- `class-aips-settings-ajax.php` `ajax_test_connection()` — `generate_text()` is intentional (just a connectivity test returning a few words).
- `class-aips-ai-assistance-service.php` — returns plain text suggestions, not JSON.

---

## Verification

1. **Research page (all providers):** Run a topic research request. Log should show "AI provider JSON response: Array" with no "Provider requested text-based JSON fallback" or retry warnings.
2. **Author suggestions:** Trigger author persona generation. Verify 12-field objects in the result.
3. **Topic generation for an author:** Trigger the author topics generator. Verify `title/score/keywords` shape in `parse_json_topics()` output.
4. **Gap analysis / content auditor:** Run a gap analysis. Verify `missing_topic/priority/reason/search_intent` fields in result.
5. **Campaign AI generation:** Trigger the AI campaign builder. Verify single-object response with no `$response[0]` needed.
6. **Dev tools scaffold:** Generate a voice/article-structure/template scaffold. Verify clean JSON returned without manual parse.
7. **Fallback still works:** Temporarily set provider to Meow (if available) and run research. The concise field-hint line in the prompt should still produce parseable JSON via `generate_json_from_text()`.
